File: targets/csharp-aspnet/Data/UserRepository.cs
using System.Data.SqlClient;
using BadDemo.Config;

namespace BadDemo.Data
{
    /// <summary>Raw ADO.NET data access.</summary>
    public class UserRepository
    {
        private SqlConnection Connect()
        {
            return new SqlConnection(AppConfig.ConnectionString);
        }

        /// <summary>Executes a SQL string provided by callers.</summary>
        public SqlDataReader QueryUnsafe(string sql)
        {
            var conn = Connect();
            conn.Open();
            var cmd = new SqlCommand(sql, conn);
            return cmd.ExecuteReader();
        }

        public void ExecuteUnsafe(string sql)
        {
            var conn = Connect();
            conn.Open();
            var cmd = new SqlCommand(sql, conn);
            cmd.ExecuteNonQuery();
        }

        public SqlDataReader QuerySafe(string sql, string arg)
        {
            var conn = Connect();
            conn.Open();
            var cmd = new SqlCommand(sql, conn);
            cmd.Parameters.AddWithValue("@p1", arg);
            return cmd.ExecuteReader();
        }
    }
}

File: targets/java-spring/data/UserRepository.java
package com.baddemo.data;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.Statement;

import com.baddemo.config.AppConfig;

/** Raw JDBC data access. The unsafe methods are the SQLi sinks. */
public class UserRepository {

    private Connection connect() throws Exception {
        return DriverManager.getConnection(AppConfig.DB_URL, AppConfig.DB_USER, AppConfig.DB_PASSWORD);
    }

    /** Sink: executes a SQL string built by callers via concatenation. */
    public ResultSet queryUnsafe(String sql) throws Exception {
        Statement stmt = connect().createStatement();
        return stmt.executeQuery(sql);
    }

    public void executeUnsafe(String sql) throws Exception {
        Statement stmt = connect().createStatement();
        stmt.executeUpdate(sql);
    }

    public ResultSet querySafe(String sql, String arg) throws Exception {
        PreparedStatement ps = connect().prepareStatement(sql);
        ps.setString(1, arg);
        return ps.executeQuery();
    }
}
